# Taller #2 — Analisis de Datos
**Realizado por: Julián Alzate Aristizábal**



## 1. Selección del Dataset

**Dataset:** El dataset es de movimiento GPS/telemetría de osos negros (*Ursus americanus*), registrado cada hora junto con variables ambientales (temperatura, NDVI, duración del día, altitud solar) y de comportamiento.

**Fuente del Dataset:** Hice una busqueda por internet y tome los datos de la página Dryad (https://datadryad.org) — *La data viene del siguiente repo: [movement/ecology dataset de osos negros]*
DOI: [10.5061/dryad.d51c5b0fd](https://datadryad.org/dataset/doi%3A10.5061/dryad.d51c5b0fd)


**Herramienta seleccionada:** Plotly (Python), sobre un notebook Jupyter (`.ipynb`).

## Base de datos para ejecutar comandos del taller

In [6]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [7]:
data = pd.read_csv(r'C:\Users\HP\Desktop\Fund_Data_Analysis\Taller_Analisis_Data\bear.csv')
data.head()

,id,Sex,age,movem_60,timestamp,season,reprod,move_status,cor_hour,altitude_dg,t_period,index,tmx,dayle,study,year,ndvi
0,1,1,2.0,0.014142,2006-05-26T03:00:00Z,1,1.0,resident,3,-21.597845,nocturnal,0.125723,19.37,14.830628,1,2,0.37345
1,1,1,2.0,0.057070,2006-05-26T04:00:00Z,1,1.0,resident,4,-14.976951,crepuscular,0.125723,19.37,14.830640,1,2,0.37345
2,1,1,2.0,0.056436,2006-05-26T05:00:00Z,1,1.0,resident,5,-6.597630,crepuscular,0.125723,19.37,14.830691,1,2,0.37345
3,1,1,2.0,0.024515,2006-05-26T06:00:00Z,1,1.0,resident,6,3.022102,crepuscular,0.125723,19.37,14.830640,1,2,0.37345
4,1,1,2.0,0.031000,2006-05-26T07:00:00Z,1,1.0,resident,7,13.474819,crepuscular,0.125723,19.37,14.830662,1,2,0.37345


## 2. Selección e investigación de la herramienta 

La herramienta que escogí fue Plotly

**a. ¿Qué es y cuál es su propósito principal?**

Plotly es una librería de Python (también existe en otros lenguajes como R y JavaScript) que sirve para crear gráficos interactivos y facilitar la visualización y exploración de datos.

**b. ¿Es gratuita, de pago o ambas?**

Es gratuita para crear gráficos y usarla en Python (Licencia de MIT). Existe además un servicio comercial para hosting, colaboración en equipo y despliegue empresarial, que sí es de pago.

**c. ¿Qué tipos de fuentes de datos permite utilizar?**

Puede trabajar con datos de Excel, CSV, bases de datos SQL, JSON, APIs, entre otros, siempre que podamos cargarlos en Python.

**d. Principales ventajas**

- Gráficos son interactivos.
- Se puede usar directamente con Pandas.
- Tiene muchos tipos de gráficos y mapas.
- Permite personalizar los títulos, etiquetas, tamaños y apariencia de los gráficos.
- Control total sobre transformación y modelado de datos porque todo ocurre en Python/Pandas.
- Escala naturalmente a una app web si el proyecto crece.
- Los gráficos se pueden exportar/gauradar y compartir en formatos como HTML e imagen.

**e. Principales limitaciones**

- Requiere saber programar (a diferencia de Tableau, Qlik o Power BI) y tener conocimientos básicos de python.
- Permite visualizar los datos, pero no la limpieza. Toca usar pandas u otras herramientas.
- Para dashboards con muchos filtros cruzados, requiere más esfuerzo de desarrollo que una herramienta de BI dedicada.
- Para datasets muy grandes puede volverse lento en el navegador a la hora de graficar.
- No analiza automáticamente los dato, se debe decidir qué variables usar y qué tipo de gráfico es adecuado.

**f. ¿En qué proyectos sería apropiado usarla?**

Es ideal para proyectos donde se necesite analizar y visualizar datos de forma interactiva y de manera exploratoria. Es útil para explorar datos, crear gráficos rápidamente, realizar informes y presentar resultados de una manera más clara. También es una buena opción en ciencia de datos y machine learning para visulizar el entrenamiento y/o resultados de un modelo.

## 3. Comprensión del dataset

In [8]:
print(f"La base de datos presenta las siguientes dimnensiones: {data.shape}")
data.info()

La base de datos presenta las siguientes dimnensiones: (379676, 17)
<class 'pandas.DataFrame'>
RangeIndex: 379676 entries, 0 to 379675
Data columns (total 17 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   id           379676 non-null  int64  
 1   Sex          379676 non-null  int64  
 2   age          378347 non-null  float64
 3   movem_60     379676 non-null  float64
 4   timestamp    379676 non-null  str    
 5   season       379676 non-null  int64  
 6   reprod       329376 non-null  float64
 7   move_status  379676 non-null  str    
 8   cor_hour     379676 non-null  int64  
 9   altitude_dg  379676 non-null  float64
 10  t_period     379676 non-null  str    
 11  index        379676 non-null  float64
 12  tmx          379676 non-null  float64
 13  dayle        379676 non-null  float64
 14  study        379676 non-null  int64  
 15  year         379676 non-null  int64  
 16  ndvi         379676 non-null  float64
dtypes: float64

**¿Qué información contiene el dataset?**

Cada fila del dataset es una posición GPS registrada por hora para un oso individual (de 6 poblaciones de osos de 11 países), junto con variables de comportamiento (estado de movimiento, período del día) y variables ambientales asociadas a ese momento y lugar (temperatura máxima, duración del día, altitud solar, índice de vegetación NDVI).

**Columnas del dataset:**
- `id`: identificador del oso.
- `Sex`: sexo codificado en binario donde, 1=female, 2=male.
- `age`: edad del oso codificado de la siguiente manera, 1=subadult (3/4 years old), 2=adult (>= 5 years old).
- `movem_60`: velocidad de movimiento en la última hora (Km/h).
- `timestamp`: fecha y hora UTC del registro.
- `season`: estación codificada entre 1 a 3, 1=spring, 2=summer, 3=fall.
- `reprod`: estado reproductivo codificado entre 1 a 3, 1=lone female, 2=male, 3=female accompanied by young
- `move_status`: estado de movimiento categórico si el oso es residente o migrante.
- `cor_hour`: hora del día corregida (1–24h).
- `altitude_dg`: altitud solar en grados (negativa de noche, positiva de día).
- `t_period`: período del día — `nocturnal`, `crepuscular`, `diurnal`.
- `index`: un índice numérico asociado al registro (probablemente de calidad de la señal GPS).
- `tmx`: temperatura máxima del día (°C).
- `dayle`: duración del día (horas de luz, "day length").
- `study`, `year`: identificadores del estudio y año.
- `ndvi`: índice de vegetación (NDVI), proxy de disponibilidad de alimento/vegetación.

In [11]:
print('Registros y variables:', data.shape[0], 'filas x', data.shape[1], 'columnas')

Registros y variables: 379676 filas x 17 columnas


**Variables más importantes para el análisis:** 
- `movem_60` la velocidad de desplazamiento del oso, que es la variable más importante del dataset de esta se podria decir que se usa para el analissi del repositorio.
-  `timestamp` / `cor_hour` son importantes para ver las horas en las que el oso se desplazó
- `t_period` la categoría en la que se desplaza el oso, podriamos evidenciar si el oso tiene más desplazamiento nocturno o diurno.
- `tmx`, `ndvi`, `dayle` y `altitude_dg` (variables ambientales que podrían explicar el movimiento).
- `season` para comparar temporadas.

In [12]:
# Valores nulos
print('Nulos por columna:')
print(data.isnull().sum())
print()
# Duplicados
print('Filas duplicadas:', data.duplicated().sum())

Nulos por columna:
id                 0
Sex                0
age             1329
movem_60           0
timestamp          0
season             0
reprod         50300
move_status        0
cor_hour           0
altitude_dg        0
t_period           0
index              0
tmx                0
dayle              0
study              0
year               0
ndvi               0
dtype: int64

Filas duplicadas: 0


In [16]:
(data.isnull().sum()/data.shape[0]).round(3)*100

id              0.0
Sex             0.0
age             0.4
movem_60        0.0
timestamp       0.0
season          0.0
reprod         13.2
move_status     0.0
cor_hour        0.0
altitude_dg     0.0
t_period        0.0
index           0.0
tmx             0.0
dayle           0.0
study           0.0
year            0.0
ndvi            0.0
dtype: float64

**El dataset no presenta duplicados, sin embargo, hay 2 variables que contienen nulos en sus registros:**
- `Age` (edad) = presenta 1329 datos nulos, que representa el 0.4% de los registros totales
- `Reprod` (estado reproductivo) = presenta 50300 datos nulos, que representa el 13,2% de los registros.

In [23]:
variables_interes = ['movem_60', 'tmx', 'ndvi', 'dayle', 'altitude_dg', 'season']

for col in variables_interes:
    q1, q3 = data[col].quantile([0.25, 0.75])
    iqr = q3 - q1
    lower, upper = q1 - 1.5*iqr, q3 + 1.5*iqr
    outliers = data[(data[col] < lower) | (data[col] > upper)]

    print(f'--- {col} ---')
    print(f'Rango intercuartílico normal: [{lower:.3f}, {upper:.3f}]')
    print(f'Registros atípicos: {len(outliers)} de {len(data)} ({len(outliers)/len(data):.1%})')
    display(outliers[[col]].head())
    print()

--- movem_60 ---
Rango intercuartílico normal: [-0.496, 0.860]
Registros atípicos: 42084 de 379676 (11.1%)


,movem_60
98,1.245490
154,1.119133
162,0.984273
281,1.019084
295,1.167244



--- tmx ---
Rango intercuartílico normal: [-1.995, 38.445]
Registros atípicos: 1674 de 379676 (0.4%)


,tmx
3342,-3.28
3344,-3.52
3345,-3.52
3346,-3.22
3347,-3.22



--- ndvi ---
Rango intercuartílico normal: [0.335, 1.042]
Registros atípicos: 15940 de 379676 (4.2%)


,ndvi
40,0.2428
41,0.2428
42,0.2428
43,0.2428
44,0.2428



--- dayle ---
Rango intercuartílico normal: [8.053, 19.873]
Registros atípicos: 5609 de 379676 (1.5%)


,dayle
266363,7.984196
266364,7.984166
266365,7.984466
266366,7.984380
266367,7.984224



--- altitude_dg ---
Rango intercuartílico normal: [-89.184, 106.791]
Registros atípicos: 0 de 379676 (0.0%)


,altitude_dg



--- season ---
Rango intercuartílico normal: [0.500, 4.500]
Registros atípicos: 0 de 379676 (0.0%)


,season


**De los variables importante de análisis se encontraron que 4 de 6 presentan datos atipicos distribuidos de la siguiente forma:**
- `movem_60`: contiene 42084 de 379676 datos son atipicos que representa el 11.1% de los regristros
- `tmx`: contiene 1674 de 379676 datos son atipicos que representa el 0.4%% de los regristros
- `ndvi`: contiene 15940 de 379676 datos son atipicos que representa el 4.2% de los regristros
- `dayle`: contiene 5609 de 379676 datos son atipicos que representa el 1.5% de los regristros

**Como preguntas de análisis se pueden plantear:**
1. ¿Cómo varía el movimiento promedio del oso (`movem_60`) según el período del día (nocturno, crepuscular, diurno)?
2. ¿Cómo evoluciona el movimiento del oso a lo largo de los 12 días de seguimiento?
3. ¿Existe relación entre el movimiento y las variables ambientales (temperatura máxima, NDVI, duración del día)?
4. ¿Cómo se distribuyen los registros entre las distintas categorías de período del día y estaciones, y cambia el patrón de movimiento entre la estación 1 y la estación 2?

## 4. Preparación de los datos